### ISSUE TITLE
sh-shellcheck ignores # shellcheck shell=dash option

### ISSUE THREAD

**Issue by joudinet** [flycheck/flycheck#2093]
Thank you for taking the time to report an issue and improve Flycheck.  This template is for **actual bugs you observed**.  If you have trouble setting up Flycheck, or if you have a question, please use the relevant issue template instead.

## Checklist

- [X] I have checked existing issues for potential duplicates before creating this one.
- [X] I have read the [Troubleshooting guide][].

## Bug description

I have a shell script that runs on embedded devices, where the sh command is actually busybox and /bin/dash does not exist. The shell script starts then with /bin/sh shebang and the next line tells shellcheck the shell is actually dash, to not report issues like for example SC3043:
#! /bin/sh
# shellcheck shell=dash

## Steps to reproduce

Create a the following script and open it with emacs:
```
#! /bin/sh -e
# shellcheck shell=dash

foo() {
  local bar="$1"
 ...
}
```

- [X] I have read https://emacs.stackexchange.com/questions/28429/how-do-i-troubleshoot-emacs-problems

## Expected behavior

The SC3043 error should not be reported here, because the actual shell is dash, as specified with the option for shellcheck.


## System configuration

```
Syntax checkers for buffer validate_image in sh-mode:                                                         
                                                                                                              
First checker to run:                                                                                         
                                                                                                              
  sh-posix-dash                                                                                               
    - may enable:    yes                                                                                      
    - may run:       t                                                                                        
    - executable:    Found at /usr/bin/dash                                                                   
    - next checkers: sh-shellcheck                                                                            
                                                                                                              
Checkers that may run as part of the first checker's chain:                                                   
                                                                                                              
  sh-shellcheck                                                                                               
    - may enable:         yes                                                                                 
    - may run:            t                                                                                   
    - executable:         Found at /usr/bin/shellcheck                                                        
    - Shell sh supported: yes                                                                                 
                                                                                                              
Checkers that could run if selected:                                                                          
                                                                                                              
  sh-posix-bash  select                                                                                       
    - may enable:    yes                                                                                      
    - may run:       t                                                                                        
    - executable:    Found at /usr/bin/bash                                                                   
    - next checkers: sh-shellcheck                                                                            
                                                                                                              
Checkers that are compatible with this mode, but will not run until properly configured:
```

Emacs configuration:
- [X] Plain Emacs / Custom configuration
- [ ] Spacemacs
- [ ] Doom Emacs
- [ ] Other shared configuration

**Comment by bbatsov** [comment]
This is solved by the `flycheck-shellcheck-infer-shell` option. When set to non-nil, Flycheck does not pass `--shell` to ShellCheck, allowing it to respect in-file directives like `# shellcheck shell=dash`.

```elisp
(setq flycheck-shellcheck-infer-shell t)
```
